src/nbl/video/ILogicalDevice.cpp

```diff
@@ -7,45 +7,52 @@
 using namespace nbl;
 using namespace nbl::video;
 
-static void debloatShaders(const asset::ISPIRVDebloater& debloater, std::span<const asset::IPipelineBase::SShaderSpecInfo> shaderSpecs, core::vector<core::smart_refctd_ptr<const asset::IShader>>& outShaders, asset::IPipelineBase::SShaderSpecInfo* outShaderSpecInfos, system::logger_opt_ptr logger = nullptr)
+class SpirvDebloatTask
 {
-    using EntryPoints = core::set<asset::ISPIRVDebloater::EntryPoint>;
-    core::map<const asset::IShader*, EntryPoints> entryPointsMap;
-
-    // collect all entry points first before we debloat
-    for (const auto& shaderSpec : shaderSpecs) {
-        const auto* shader = shaderSpec.shader;
-        auto it = entryPointsMap.find(shader);
-        if (it == entryPointsMap.end() || it->first != shader)
-            it = entryPointsMap.emplace_hint(it, shader, EntryPoints());
-        it->second.insert({ .name = shaderSpec.entryPoint, .stage = shaderSpec.stage });
-    }
+    public:
+      using EntryPoints = core::set<asset::ISPIRVDebloater::EntryPoint>;
 
-    core::map<const asset::IShader*, const asset::IShader*> debloatedShaders;
-    for (const auto& shaderSpec: shaderSpecs)
-    {
-        const auto* shader = shaderSpec.shader;
-        const auto& entryPoints = entryPointsMap[shader];
+        SpirvDebloatTask(asset::ISPIRVDebloater* debloater, system::logger_opt_ptr logger) : m_debloater(debloater), m_logger(logger)
+        {
+          
+        }
 
-        auto debloatedShaderSpec = shaderSpec;
-        if (shader != nullptr)
+        void insertEntryPoint(const IGPUPipelineBase::SShaderSpecInfo& shaderSpec, hlsl::ShaderStage stage)
         {
-            if (!debloatedShaders.contains(shader))
-            {
-                const auto outShadersData = outShaders.data();
-                outShaders.push_back(debloater.debloat(shader, entryPoints, logger));
-                assert(outShadersData == outShaders.data());
-                debloatedShaders.emplace(shader, outShaders.back().get());
-            }
-            const auto debloatedShader = debloatedShaders[shader];
-            debloatedShaderSpec.shader = debloatedShader;
+            const auto* shader = shaderSpec.shader;
+            auto it = m_entryPointsMap.find(shader);
+            if (it == m_entryPointsMap.end() || it->first != shader)
+                it = m_entryPointsMap.emplace_hint(it, shader, EntryPoints());
+            it->second.insert({ .name = shaderSpec.entryPoint, .stage = stage });
         }
-        *outShaderSpecInfos = debloatedShaderSpec;
 
-        outShaderSpecInfos++;
-    }
+        IGPUPipelineBase::SShaderSpecInfo debloat(const IGPUPipelineBase::SShaderSpecInfo& shaderSpec, core::vector<core::smart_refctd_ptr<const asset::IShader>>& outShaders)
+        {
+            const auto* shader = shaderSpec.shader;
+            const auto& entryPoints = m_entryPointsMap[shader];
```

Comment:
you may want to use `find` and assert that you've found (basically that someone called `insertEntryPoint` on the `shaderSpec` before)